### ISSUE TITLE
Add hints to random access availability to archives

### ISSUE THREAD

**Issue by vpenades** [adamhathcock/sharpcompress#708]
Following #707

I've noticed you really don't know if you need to read sequentially or you can access randomly until you have opened the archive... correct me if I'm wrong:

Plain .TAR files can actually be read randomly.... but tar archives inside a gzip archive can't

So, if that's the case, the "SupportsRandomAccess" property should go into the IArchive, not in the IArchiveFactory interface, right?

Also, wouldn't the IsSolid property would be enough for this? (in that case, IsSolid should be true for .tar.gz files)

If IsSolid serves a different purpose, then, IArchive definitely needs a SupportsRandomAccess

Any thoughts?

**Comment by adamhathcock** [comment]
IsSolid means a specific thing for RAR.

On Streams there's IsSeekable which basically means what we want but we need it on IArchive.  An alternative is just not to have IArchive interfaces for non-seekable situations and tell people to use IReader

**Comment by vpenades** [comment]
Got it...  I'll do a new PR with that knowledge

**Comment by vpenades** [comment]
Hmm.. having a hard time exposing the IsSeekable to IArchive ....  opening a tar.gz  reports IsSeekable to true in all the streams I can see around.

**Comment by adamhathcock** [comment]
IReaderFactory has that Open I think.

You don't want to put stream's IsSeekable on IArchive.  You want to return true/false based on the archive/compression format.  File streams are always seekable but decompressing files is usually not.  Zip/Rar has individual files compressed so can seek.  TarGz/TarBz are one continuous compression so they're not seekable.

**Comment by vpenades** [comment]
yes, it's ReaderFactory the one that has that Open...  so I think an IReaderFactory interface is needed, in the same way that I introduced IArchiveFactory... so the spaguetty code can be removed and new readers can be registered

**Comment by adamhathcock** [comment]
that's not bad idea to just have singular factory classes or something to consolidate things

**Comment by vpenades** [comment]
So, SevenZip doesn't have an implementation in the readers factory?

**Comment by vpenades** [comment]
Added a PR: #709

**Comment by adamhathcock** [comment]
> So, SevenZip doesn't have an implementation in the readers factory? is there a reason for it?

This is because 7Zip requires random access to a file from my memory.  The streams need to seek around to properly find headers and decompress the streams in the format.  Readers only work for non-seekable streams.

**Comment by vpenades** [comment]
@adamhathcock expanding this topic a bit further:  which would be the recomended way to open archives in a generic way?

What I'm trying to achieve is to traverse a number of directories, containing all sorts of archives (zip,rar, 7z, etc) open them and scan their content.

**Comment by adamhathcock** [comment]
You'll have to implement that yourself if you can't guarantee Reader.  It's beyond the scope of the library. 

I've been away for personal reasons.
